### ISSUE TITLE
rm(linkToDir) will wipe dir content

### ISSUE THREAD

**Issue by arberg** [lihaoyi/os-lib#4]
See tests below, and conversation https://gitter.im/lihaoyi/Ammonite (24-11-2018), and see same issue reported in Ammonite: https://github.com/lihaoyi/Ammonite/issues/885

I created this issue, since its now a bug in os-lib, now that the codebase has been moved.

According to my tests and code-review, the rm(dir) does not recursively follows links and wipes dir-content, its just rm(link) that's broken. It would be valuable though to ensure it stays that way by incorporating a test like below.

	// https://github.com/lihaoyi/os-lib/commits/master
	object MyTests {
		val ammTestDir=Path("/mnt/user/home/alex/AmmoniteTestRm/main")
		def setupTestRm() {
			rm! ammTestDir
			mkdir! ammTestDir/'sub1
			mkdir! ammTestDir/'sub2
			write.over(ammTestDir/'file, "myContent")
			write.over(ammTestDir/'sub1/'file1, "myContent1")
			write.over(ammTestDir/'sub2/'file2, "myContent2")
			ln.s(ammTestDir, ammTestDir/'linkToRootMain)
			ln.s(ammTestDir/'sub1, ammTestDir/'linkToSub1)
			ln.s(ammTestDir/'sub2, ammTestDir/'sub1/'linkToSub2ParentSibling)
		}

		def test_rm_link_to_sub1_from_main() {
			try {
				rm! ammTestDir/'linkToSub1
				assert(exists(ammTestDir/'sub1))
				assert(exists(ammTestDir/'sub1/'file1))
				assert(exists(ammTestDir/'sub2/'file2)) // Might be deleted because sub1 contains link to sub2 dir
			} catch {
				case e : AssertionError => println("!!!!!!!!!!!!! Failed Test: sub1 file deleted when deleting link to dir. rm followed link and deleted content")
				case e => println("!!!!!!!!!!!!! rm error: "+e.getMessage()+e)
			}
		}
		def test_rm_dir_that_contains_a_link_to_other_dir() {
			try {
				rm! ammTestDir/'sub1
				assert(exists(ammTestDir/'sub2/'file2))
			} catch {
				case e : AssertionError => println("!!!!!!!!!!!!! Failed Test: sub2 file deleted when trying to remove sub1 dir, it should not have followed the link to sub2 and deleted its content")
				case e => println("!!!!!!!!!!!!! rm error: "+e.getMessage()+e)
			}

		}
		def runSuite() {
			setupTestRm()
			test_rm_link_to_sub1_from_main()
			setupTestRm()
			test_rm_dir_that_contains_a_link_to_other_dir()
		}
	}

**Comment by lihaoyi** [comment]
should be fixed in 0.2.3 https://github.com/lihaoyi/os-lib/commit/ddb562f8794169b19825be94467b798567673edb

**Comment by arberg** [comment]
Hi Li, 

I've tested 1.6.0 and indeed it is now fixed.

It great you've added tests to ensure it doesn't resurface. 

I noticed in my review of the [tests ](https://github.com/lihaoyi/os-lib/commit/ddb562f8794169b19825be94467b798567673edb) that it would help 'document' the test-code if you checked that the link you remove actually exists just before the deletion. It will both document the code and work as a test of the test, without unduly complicating the code in my view. 

Best Alex
